Locals

yorinasub17 opened the pull request:
This implements the concept of `locals` in terragrunt config. This can be used to bind expressions to names for reusing within the config, so that you can DRY complex expressions within the config.

TODO:
- [x] Add more testing

jfunnell commented:
Wow awesome, this is an excellent improvement if it works (might test it today..). 

Like you said, I don't really see how this could have helped with issue #752 and that is fine.. I think the overall best solution to clean up some of the shortcomings of terragrunt is to implement the entire feature set in #759, which would probably require a hefty chunk of code/functionality to change.  
In short, I think this is an acceptable stopgap for now. Just my 2c

brikis98 commented on README.md:
```diff
@@ -1251,6 +1252,26 @@ Note that Terragrunt will respect any `TF_VAR_xxx` variables you've manually set
 
+### Locals
+
+You can use locals to bind a name to values and expressions that you would like to create a reference for within your
```
```suggestion
You can use locals to bind a name to an expression, so you can reuse that expression without having to repeat it multiple times (keeping your Terragrunt configuration DRY).
```

brikis98 commented on README.md:
````diff
@@ -1251,6 +1252,26 @@ Note that Terragrunt will respect any `TF_VAR_xxx` variables you've manually set
+}
+```
+
+You can use any valid terragrunt expression in the `locals` configuration.
````
Prepare yourself for the "how do I reuse locals from another file?" question in 3... 2... 1... :-D

Perhaps tackle that head on and add a section on reusing/merging `yaml` files for now? And when we add `get_input` and `get_output`, we can replace it.

yorinasub17 replied:
Ok I can add that in.

FWIW, I was thinking about proposing a similar construct, but called `globals` which can be merged with the parent in an `include` call. I think the explicit separating of `locals` from `globals` is important, especially if and when `include` becomes more complex to allow inclusion of multiple levels of terragrunt config.

apottere replied:
Do you have any immediate plans to implement `globals`?  We've just developed a need for something like that (having variables in our parent `terragrunt.hcl` that we can override from child `terragrunt.hcl`s).  While including a yaml file if it exists would work, it doesn't seem like the cleanest solution and we'd have to walk it back once there's a better way.

If you haven't started and you think it would be something that would be accepted, I don't mind creating a PR for it.

yorinasub17 replied:
> If you haven't started and you think it would be something that would be accepted, I don't mind creating a PR for it.

I don't think we have immediate plans for `globals`. I would suggest first creating an issue (with `Feature Request` in the title) describing:

- Your use case. Why do you need something like `globals`?
- Ideal solution. Is `globals` what you want, or do you want something like `get_local`, which can take another `terragrunt.hcl` config file as input, a `key` and return the value assigned to that `local` in the target file?
- Indicate whether you want to open a PR, either for the ideal solution or if that is challenging, a temporary workaround solution that works, but is better than what is available now.

brikis98 commented on config/config.go:
```diff
@@ -299,24 +312,48 @@ func ParseConfigString(configString string, terragruntOptions *options.Terragrun
+// the config. For consistency, `locals` in the call to `decodeHcl` is always assumed to be nil.
+func decodeAsTerragruntLocals(file *hcl.File, filename string, terragruntOptions *options.TerragruntOptions) (*terragruntLocals, error) {
+	terragruntLocals := terragruntLocals{}
+	err := decodeHcl(file, filename, &terragruntLocals, terragruntOptions, nil, nil)
```
What happens if a local references another local?

yorinasub17 replied:
Ah I didn't think about this possibility. I think currently it will panic or raise a syntax error, but it would be nice if it can be recursive. Let me do a quick spike to see how hard it is to implement this.

yorinasub17 commented:
UPDATE:

I made a huge refactor of the implementation to convert the `locals` attribute to a block. This has the advantage of being able to reference other locals in the `locals` block, but at the expense of being unable to merge directly from `yamldecode`. I think this is an ok tradeoff considering that you can still import yamls as `locals`, only that it can't be merged at the top level (see the new docs I wrote).

Note that I am still optimizing the logs and error handling, so not quite ready to merge yet but I reached the point where it would be good to get some feedback on the new approach.

yorinasub17 commented:
Ok @brikis98 this is ready for re-review given the new implementation.

brikis98 commented on README.md:
````diff
@@ -1253,6 +1254,79 @@ Note that Terragrunt will respect any `TF_VAR_xxx` variables you've manually set
+}
+```
+
+##### Including globally defined locals
````
Thx for adding 👍

brikis98 commented on config/config.go:
```diff
@@ -45,6 +47,15 @@ type terragruntConfigFile struct {
 	Skip           *bool                  `hcl:"skip,attr"`
 	IamRole        *string                `hcl:"iam_role,attr"`
+
+	// This should be ignored
```
Why ignored?

yorinasub17 replied:
This needs to be skipped here because `hcl` can't decode it correctly when it self references `locals`. It is not necessary to decode it into this struct here because it is evaluated in a separate loop.

However, we don't want to completely omit the existence of it because we want a syntax error if we have blocks that terragrunt doesn't expect (e.g typos). Hence it is added here as an attribute but the struct won't attempt to parse anything in it.

brikis98 replied:
Perhaps expand the comment a bit to explain this?

yorinasub17 replied:
Already did: https://github.com/gruntwork-io/terragrunt/pull/802/files#diff-3baf47c64847a8fb8aaa8cc2e088513bR51

brikis98 commented on config/config.go:
```diff
@@ -299,24 +318,37 @@ func ParseConfigString(configString string, terragruntOptions *options.Terragrun
 func decodeAsTerragruntInclude(file *hcl.File, filename string, terragruntOptions *options.TerragruntOptions) (*terragruntInclude, error) {
 	terragruntInclude := terragruntInclude{}
-	err := decodeHcl(file, filename, &terragruntInclude, terragruntOptions, nil)
+	err := decodeHcl(file, filename, &terragruntInclude, terragruntOptions, nil, nil)
```
An `include` can't have its own locals?

yorinasub17 replied:
Ah good point! Updated the evaluation order so it can be there.

brikis98 commented on config/locals.go:
```diff
@@ -0,0 +1,313 @@
+	}
+	if localsBlock == nil {
+		// No locals block referenced in the file
+		terragruntOptions.Logger.Printf("Did not find any locals block: skipping evaluation.")
```
Do we need to log this?

yorinasub17 replied:
Implemented a crude version of log levels. This is actually quite useful to debug issues with locals and I expect we will need it once people start using it.

brikis98 commented on config/locals.go:
```diff
@@ -0,0 +1,313 @@
+		return nil, nil
+	}
+
+	terragruntOptions.Logger.Printf("Found locals block: evaluating the expressions.")
```
Same here... Terragrunt doesn't currently support log levels, so the log statements here could be quite noisy...

yorinasub17 replied:
See above re:log levels.

brikis98 commented on config/locals.go:
```diff
@@ -0,0 +1,313 @@
+	// further.
+	evaluatedLocals := map[string]cty.Value{}
+	evaluated := true
+	for len(locals) > 0 && evaluated {
```
Should we add a simple counter and ensure it stays under, say, 1000? Could be a nice safety valve in case of a bug in the code that leads to an infinite loop.

Actually, on that note, what if the code has:

```hcl
locals {
  a = local.b
  b = local.a
}
```

yorinasub17 replied:
This will fail because neither can be evaluated in the first loop, so `evaluated` will be `false`.

yorinasub17 replied:
> Should we add a simple counter and ensure it stays under, say, 1000? Could be a nice safety valve in case of a bug in the code that leads to an infinite loop.

Implemented.

brikis98 commented on config/locals.go:
```diff
@@ -0,0 +1,313 @@
+// - the updated map of evaluated locals after this attempt
+// - whether or not any locals were evaluated in this attempt
+// - any errors from the evaluation
+func attemptEvaluateLocals(
```
Yikes, this is quite complicated. I have some vague memory that the HCL parser has native support for partial evaluation specifically for this use case, but can't find it in the code base quickly...

yorinasub17 replied:
I searched but the only partial evaluation I could find was at the top file level, and not at the individual block level where we have expressions.

brikis98 commented on config/locals_test.go:
```diff
@@ -0,0 +1,83 @@
+  // Multiple references
+  x = 1
+  y = 2
+  z = local.x + local.y
```
Consider testing a few corner cases:

* Reference to a `local` that isn't defined in a `locals { ... }` block
* Reference to a `local` that isn't defined elsewhere in the code
* Cyclical references
* Several layers of references (a = local.b, b = local.c, c = local.d, etc)

yorinasub17 replied:
Done in a107c89

brikis98 commented on util/logger.go:
```diff
@@ -19,3 +23,22 @@ func CreateLoggerWithWriter(writer io.Writer, prefix string) *log.Logger {
+	if os.Getenv("TG_LOG") == "DEBUG" {
+		logger.Printf(fmtString, fmtArgs...)
+	}
+}
```
Ah, clever 👍

yorinasub17 commented:
UPDATE a107c89 :

- Added a bunch more test cases
- Fixed a bug where we parsed multiple locals block, but silently ignored all but the first one defined. Now we throw an error if we detect multiple locals blocks.
- Fixed a case where we weren't properly writing out diagnostics information.

With that, if the build passes, I think we are ready to merge + release!

brikis98 commented:
LGTM!

yorinasub17 commented:
Ok merging and releasing. Thanks for the review!
